File: buildsrht/manifest.py
from enum import Enum
import os
import re
import subprocess
import uuid
import yaml

class TriggerAction(Enum):
    email = 'email'
    webhook = 'webhook'

class TriggerCondition(Enum):
    success = 'success'
    failure = 'failure'
    always = 'always'

class Trigger:
    def __init__(self, yml):
        if not isinstance(yml, dict):
            raise Exception("Expected trigger to be a dict")
        self.action = TriggerAction(yml["action"])
        self.condition = TriggerCondition(yml["condition"])
        self.attrs = {
            key: yml[key] for key in yml.keys()
                if key not in ["action", "condition"]
        }

    def to_dict(self):
        attrs = self.attrs
        attrs.update({
            "action": self.action.value,
            "condition": self.condition.value,
        })
        return attrs

class Task:
    def __init__(self, yml):
        if not isinstance(yml, dict) or len(yml) != 1:
            raise Exception("Expected task to be a string: string")
        for key in yml:
            if not isinstance(key, str) or not isinstance(yml[key], str):
                raise Exception("Expected task to be a string: string")
            self.name = key
            self.script = yml[key].strip()
        if not re.match(r"^[a-z0-9_-]+$", self.name) or len(self.name) > 128:
            raise Exception(f"Task name '{self.name}' is invalid (must be all lowercase letters, " +
                "numbers, underscores, and dashes, and <=128 characters)")

    def __repr__(self):
        return "<Task {}>".format(self.name)

class Manifest:
    def __init__(self, yml):
        self.yaml = yml
        image = self.yaml.get("image")
        arch = self.yaml.get("arch")
        packages = self.yaml.get("packages")
        repos = self.yaml.get("repositories")
        sources = self.yaml.get("sources")
        env = self.yaml.get("environment")
        secrets = self.yaml.get("secrets")
        shell = self.yaml.get("shell")
        artifacts = self.yaml.get("artifacts")
        oauth = self.yaml.get("oauth")
        if not image:
            raise Exception("Missing image in manifest")
        if not isinstance(image, str):
            raise Exception("Expected image to be a string")
        if packages:
            if not isinstance(packages, list) or not all([isinstance(p, str) for p in packages]):
                raise Exception("Expected packages to be a string array")
        if repos:
            if not isinstance(repos, dict):
                raise Exception("Expected repositories to be a dict")
            for repo in repos:
                if not isinstance(repos[repo], str):
                    raise Exception("Expected url for repository {}".format(repo))
        if sources:
            if not isinstance(sources, list) or not all([isinstance(s, str) for s in sources]):
                raise Exception("Expected sources to be a string array")
        if env:
            if not isinstance(env, dict):
                raise Exception("Expected environment to be a dictionary")
        if secrets:
            if not isinstance(secrets, list) or not all([isinstance(s, str) for s in secrets]):
                raise Exception("Expected secrets to be a UUID array")
            # Will throw exception on invalid UUIDs as well
            secrets = list(map(uuid.UUID, secrets))
        if shell is not None and not isinstance(shell, bool):
            raise Exception("Expected shell to be a boolean")
        if artifacts is not None and (
                not isinstance(artifacts, list) or
                any(not isinstance(p, str) or p == "" for p in artifacts) or
                len(set(os.path.basename(p) for p in artifacts)) != len(artifacts)):
            raise Exception("Expected artifacts to be a list of unique file paths")
        if oauth is not None and not isinstance(oauth, str):
            raise Exception("Expected oauth to be a string")
        self.image = image
        self.arch = arch
        self.packages = packages
        self.repos = repos
        self.sources = sources
        self.environment = env
        self.secrets = secrets
        self.shell = shell
        self.artifacts = artifacts
        self.oauth = oauth
        tasks = self.yaml.get("tasks")
        if not tasks or not isinstance(tasks, list):
            if (tasks is None or tasks == []) and not self.shell:
                raise Exception("Attempted to create manifest with no tasks")
            else:
                tasks = []
        self.tasks = [Task(t) for t in tasks]
        for task in self.tasks:
            if len([t for t in self.tasks if t.name == task.name]) != 1:
                raise Exception(f"Duplicate task '{task.name}'")
        triggers = self.yaml.get("triggers")
        self.triggers = [Trigger(t) for t in triggers] if triggers else []

    def __repr__(self):
        return "<Manifest {}, {} tasks>".format(self.image, len(self.tasks))
    
    def to_dict(self):
        return {
            "image": self.image,
            "arch": self.arch,
            "packages": self.packages,
            "repositories": self.repos,
            "sources": self.sources,
            "environment": self.environment,
            "secrets": [str(s) for s in self.secrets] if self.secrets else None,
            "tasks": [{ t.name: t.script } for t in self.tasks],
            "triggers": [t.to_dict() for t in self.triggers] if any(self.triggers) else None,
            "shell": self.shell,
            "artifacts": self.artifacts,
            "oauth": self.oauth,
        }

    def to_yaml(self):
        return yaml.dump(self.to_dict())

File: api/graph/manifest.go
package graph

import (
	"fmt"
	"regexp"

	"github.com/google/uuid"
	"gopkg.in/yaml.v2"
)

var taskRE = regexp.MustCompile("^[a-z0-9_-]+$")

type Trigger struct {
	Action    string `yaml:"action",json:"action"`
	Condition string `yaml:"condition",json:"condition"`

	// Email fields
	To        *string `yaml:"to",json:"to,omitempty"`
	Cc        *string `yaml:"cc",json:"cc,omitempty"`
	InReplyTo *string `yaml:"in_reply_to",json:"in_reply_to,omitempty"`

	// Webhook fields
	Url *string `yaml:"url",json:"url"`
}

type Manifest struct {
	Arch         *string                `yaml:"arch",json:"arch,omitempty"`
	Artifacts    []string               `yaml:"artifacts",json:"artifacts,omitempty"`
	Environment  map[string]interface{} `yaml:"environment",json:"environment,omitempty"`
	Image        string                 `yaml:"image",json:"image"`
	Packages     []string               `yaml:"packages",json:"packages,omitempty"`
	Repositories map[string]string      `yaml:"repositories",json:"repositories,omitempty"`
	Secrets      []string               `yaml:"secrets",json:"secrets,omitempty"`
	Shell        bool                   `yaml:"shell",json:"shell,omitempty"`
	Sources      []string               `yaml:"sources",json:"sources,omitempty"`
	Tasks        []map[string]string    `yaml:"tasks",json:"tasks"`
	Triggers     []Trigger              `yaml:"triggers",json:"triggers,omitempty"`
	OAuth        string                 `yaml:"oauth",json:"oauth,omitempty"`
}

func LoadManifest(in string) (*Manifest, error) {
	var manifest Manifest
	err := yaml.Unmarshal([]byte(in), &manifest)
	if err != nil {
		return nil, err
	}

	if manifest.Image == "" {
		return nil, fmt.Errorf("image is required")
	}

	for _, sec := range manifest.Secrets {
		_, err := uuid.Parse(sec)
		if err != nil {
			return nil, err
		}
	}

	artset := make(map[string]interface{})
	for _, art := range manifest.Artifacts {
		if _, ok := artset[art]; ok {
			return nil, fmt.Errorf("duplicate artifact %s", art)
		}
		artset[art] = nil
	}

	if len(manifest.Tasks) == 0 && !manifest.Shell {
		return nil, fmt.Errorf("list of tasks is required")
	}

	taskset := make(map[string]interface{})
	for _, task := range manifest.Tasks {
		if len(task) != 1 {
			return nil, fmt.Errorf(`task schema is {"name": "value"} (or 'taskname: script...' as YAML)`)
		}
		var name string
		for key, _ := range task {
			name = key
			break
		}

		if _, ok := taskset[name]; ok {
			return nil, fmt.Errorf("duplicate task %s", name)
		}
		taskset[name] = nil

		if !taskRE.Match([]byte(name)) {
			return nil, fmt.Errorf("invalid task name '%s'", name)
		}
	}

	for _, trigger := range manifest.Triggers {
		switch trigger.Action {
		case "email":
			if trigger.To == nil && trigger.Cc == nil {
				return nil, fmt.Errorf("email trigger requires 'to' or 'cc'")
			}
		case "webhook":
			if trigger.Url == nil {
				return nil, fmt.Errorf("webhook trigger requires 'url'")
			}
		default:
			return nil, fmt.Errorf("unknown trigger type '%s'", trigger.Action)
		}
	}

	return &manifest, nil
}
